keps/sig-apps/20190412-pdb-for-scale-subresource.md

```diff
@@ -0,0 +1,84 @@
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Support scale subresource in PDBs
+
+## Table of Contents
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+- [Proposal](#proposal)
+  - [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+
+
+## Summary
+
+[Pod Disruption Budget (PDB)](https://kubernetes.io/docs/tasks/run-application/configure-pdb/)
+is a Kubernetes API that limits the number of pods of a collection that are down simultaneously from voluntary disruptions. PDBs allows a user to specify the allowed disruption through either min available or max unavailable number of pods. In order to support PDBs where max number of unavailable pods is set, the PDB controller needs to be able to look up the desired number of replicas. It does this by looking at the controller(s) (as specified by the owner ref) for the pods covered by the PDB. Currently only four of the basic workload controllers are supported for PDBs, namely Deployment, StatefulSet, ReplicaSet and ReplicaController. 
+
+The scale subresource allow any resource to specify its desired number of replicas and a generic way to look up this information. This document lays out a plan to use the scale subresource to allow setting PDBs for any resource implementing the scale subresource.
+
+## Motivation
+
+PDBs are an important tool to control the number of voluntary disruptions for workloads on Kubernetes. As more users start deploying custom controllers/operators based on CRDs, it is inconvenient that PDBs are not available. The scale subresource is already in use by the autoscaler and we should extend its use to PDBs.
+
+### Goals
+
+- Implement support for scale subresources in PDBs
+- Avoid major performance impact from the change
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+In the current implementation, the PDB controller identifies the workload controller by going through all the pods covered by the PDB, and for each pod, check in sequence for each of the four workload controllers. The PDB controller then looks up the desired number of replicas from the identified controller. If the set of pods covered by the PDB identfies more than one controller, that is considered an error. 
+
+The plan is to keep this approach, but check for the scale subresource if neither of the kubernetes workload controllers match. Each of the kubernetes workload controllers actually implement the scale subresource, so it is possible to only use that one. But since shared informers can not be used with the scale subresource, we want to rely on the informers for checking the kubernetes workload controllers, and only try the scale subresource if none of the them match.
```

Comment:
Maybe it's actually time to think about supporting for scale subresource in informers? It's not the first time it's hitting us...
@kubernetes/sig-api-machinery-feature-requests

Reply:
I'm not sure what you mean by scale subresource in informers... the scale subresource doesn't support watch.

Reply:
Sorry, that was my mental shortcuts. I basically mean support watch for scale subresource.
Or is it something we don't want to do because of some reason?

Reply:
There are few things here:
1. adding watch support for a subresource (so you could watch `/apis/apps/v1/namespaces/default/deployments/foo/scale`). Watch is currently a variant of list, which also isn't supported on subresources. We'd need to think through the semantics of that.
2. Opening individual watches on every scale subresource of a workload object doesn't seem like it would be great... are you suggesting watching scale subresources *across* multiple parent resources? There isn't currently an API pattern that lets you request subresources across multiple parent resources. The semantics of merging multiple subresources into a single resource are also unclear to me.

Reply:
Thanks Jordan!

> are you suggesting watching scale subresources across multiple parent resources? 

That was what I was thinking about.

Another alternative (though I guess it sounds hacky...) would be:
- ultimately, subresource just performs some actions on the whole resource
- so if we would be able to somehow extract the "subresource" logic, maybe we would be able to apply it client-side too?
- in such case, we would just be able to watch objects, and just apply subresource (e.g. GET .../scale) client side, but doing GET and manipulating returned resource

Reply:
I don't think watching subresources is necessary for this KEP--I'd prefer not to take that on unless we have a really compelling reason since (as Jordan points out) there's no clear way to express this in our api.

Reply:
Agree with Daniel, adding watch for subresources is out of scope imo.